Protobuffer generation: handle many parameters

nityanandaz opened the pull request:
## Problem that is solved

- Closes #62
- Closes #140

## Implications

- Files in _SMB/gRPC/ProtobufferBuilder_ moved to _SMB/Protobuffer_
- _ProtobufferBuilderTests_ and _~ProtobufferBuilderInternalTests~_ are merged
- `EnrichedInfo.propertyInfo` and `EnrichedInfo.propertyOffset` are merged into `Apodini.PropertyInfo`. The information scope is narrower.

## Related PRs

The feedback of #150 will implemented there.

## Reviewer Nudging

I apologize that this PR got larger that anticipated, but I had the urge to move the implementation closer together.

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Handle many parameters

nityanandaz committed:
ProtobufferBuilderTests.testServiceParametersTwo fails

nityanandaz committed:
Make PropertyInfo editable (outside of Runtime)

nityanandaz committed:
Merge ProtobufferBuilderTests

nityanandaz committed:
Refactor ProtobufferBuilderTests.buildMessage

And add Node.collectValues,
delete ProtobufferBuilder.addService/addMessage

nityanandaz committed:
SwiftLint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/ParticularType.swift
#	Tests/ApodiniTests/ProtobufferBuilderTests/ProtobufferBuilderTests.swift

nityanandaz committed:
Supplement

nityanandaz committed:
Use mangledName(of:)

nityanandaz committed:
Remove differentiation

Trade client-side "convenience" for simplicity

nityanandaz committed:
Consider GRPCParameterOptions

PSchmiedmayer committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Refactor ProtobufferBuilder.analyze

nityanandaz committed:
Add comment

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Create isAmbiguousSupportedFixedWidthInteger

Move name strategy into Model+EnrichedInfo.swift

nityanandaz committed:
Pass tests

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Supplement

nityanandaz committed:
Supplment 2

nityanandaz committed:
Refactor EndpointsTreeNode.collectAllEndpoints with `Node`

nityanandaz committed:
Rename collectAllEndpoints -> collectEndpoints

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

# Conflicts:
#	Sources/Apodini/Utilities/TypeInfo.swift

nityanandaz committed:
Apply feedback

nityanandaz committed:
SwiftLint

moritzschuell reviewed (changes requested):
For some reason, this branch's `TestWebService` does not work for me 🤔
It fails in `EnrichedInfo.swift` line 58 with the `preconditionFailure(errorDescription)`:

`Thread 1: Fatal error: couldNotGetTypeInfo(type: () -> Swift.UInt, kind: Runtime.Kind.function)`

I tried clean-building and also replacing any `Int`s or `UInt`s in the `TestWebService` with their fixed-width counterparts, but it still keeps crashing.

Tests are passing like a charm, though.

nityanandaz committed:
Fix `Parameter` has closure property

https://github.com/Apodini/Apodini/pull/131#pullrequestreview-571524716

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/Tree.swift
#	Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferInterfaceExporter.swift

nityanandaz committed:
Move files, fix headers

nityanandaz committed:
Filter Handler Parameters from all other property wrappers, e.g., Environment.

nityanandaz committed:
Move .../gRPC/ProtobufferBuilder

nityanandaz committed:
Make ProtobufferInterfaceExporter a StaticInterfaceExporter

nityanandaz committed:
Remove private ProtobufferBuilderTests extension

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/EnrichedInfo.swift

nityanandaz committed:
Buff ProtobufferInterfaceExporter 💪

Nerf ProtobufferBuilder 🚶‍♂️

nityanandaz committed:
Merge branch 'protobuffer-generation/handle-many-parameters' into protobuffer-generation/support-int-and-uint

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Protobuffer/Model+EnrichedInfo.swift
#	Sources/Apodini/Semantic Model Builder/Protobuffer/TypeInfo+NameStrategy.swift

nityanandaz committed:
Supplement

nityanandaz committed:
Merge branch 'protobuffer-generation/support-int-and-uint' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Pass tests

Previously, scalar parameter messages were treated as composite message in the handler message

nityanandaz committed:
SwiftLint

nityanandaz committed:
Words

nityanandaz committed:
Create and pass testVariableWidthInteger

nityanandaz committed:
Supplement

nityanandaz committed:
Consistency

lschlesinger commented on Sources/Apodini/Semantic Model Builder/Protobuffer/Model+EnrichedInfo.swift:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+    
+    func compatibleGenericName() throws -> String {
```
I just today did the same task for OAS: handling type names including incompatible characters, such as `<`, `>`, and `,` 😄 ! I tried to not hide the information by what these characters are replaced, through defining an enum including some constants:
```swift
enum OpenAPISchemaConstants {
    static let replaceAngleBracket = "_"
    static let replaceCommaSeparation = "-comma-"
}
```
This also allowed by to use these properties in the tests and not getting out of sync in case I want to change something! Maybe you find that useful, too.
I came up with this enum solution for constants, as this is the only format swiftlint will not complain about.

lschlesinger replied:
Additionally, I used regex, to catch cases where `X, Y` lacks the whitespace: `X,Y`. So I use:
```swift
.replacingOccurrences(of: "(\\,\\s|\\,)", with: OpenAPISchemaConstants.replaceCommaSeparation, options: .regularExpression)
```

nityanandaz replied:
Did you catch any cases were the space is actually missing? I expect this behavior to be stable 🙂

lschlesinger replied:
I thought I had seen it somewhere in the debugger... but I just checked again, you're right!

lschlesinger commented on Sources/Apodini/Utilities/Tree.swift:
```diff
@@ -141,6 +141,28 @@ extension Node {
 }
 
+extension Node {
+    func collectValues() -> Set<T> where T: Hashable {
```
I find this very useful!! And as I already sneaked a peek at that PR before, I already incorporated this in my work by simply copying this to check if it solves my task - it worked. So I am happy to find this in develop, soon 👍

lschlesinger commented on Sources/Apodini/Semantic Model Builder/Protobuffer/ProtobufferInterfaceExporter.swift:
```diff
@@ -0,0 +1,197 @@
+    }
+    
+    static func buildCompositeMessage(_ type: Any.Type) throws -> Tree<ProtobufferMessage> {
+        try EnrichedInfo.node(type)
```
I encountered that I need to do this "editing process" recursively, in order to catch the cases, where collection types are nested. The related PR #179 is not ready yet, but I think it is enough for you to identify if you might run in the same problems. The associated Issue #178 hopefully help to understand, what I try to reach in the PR.

nityanandaz replied:
I am also working/experimenting on a solution for this: https://github.com/Apodini/Apodini/blob/156cba7df7d1dbceb877cf95fd463078535770b1/Sources/Apodini/Semantic%20Model%20Builder/Model/TypeReflection/EnrichedInfo.swift#L148 on https://github.com/Apodini/Apodini/tree/refactor/EnrichedInfo.node

I will look into it 👍

lschlesinger reviewed (approved):
I find the `propertyInfo` refactoring reasonable 👍 Added comments regarding the string replacement in `compatibleType` and a link to a PR which I think is related to your `.edited()` procedure. If so, this can also be addressed in a separate PR - this one is good to go I would say 😉

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Add tree documentation

nityanandaz committed:
Revise tree documentation

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

moritzschuell reviewed (approved):
Looks good to me 👍 
This is a big step forward, thank you!

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Supplement
